hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -953,9 +956,26 @@ private void finishActiveMasterInitialization(MonitoredTask status)
     if (!waitForMetaOnline()) {
       return;
     }
+    TableDescriptor metaDescriptor = tableDescriptors.get(
+        TableName.META_TABLE_NAME);
+    final ColumnFamilyDescriptor tableFamilyDesc = metaDescriptor
+        .getColumnFamily(HConstants.TABLE_FAMILY);
+    final ColumnFamilyDescriptor replBarrierFamilyDesc =
+        metaDescriptor.getColumnFamily(HConstants.REPLICATION_BARRIER_FAMILY);
+
     this.assignmentManager.joinCluster();
     // The below depends on hbase:meta being online.
-    this.tableStateManager.start();
+    try {
+      this.tableStateManager.start();
+    } catch (NoSuchColumnFamilyException e) {
+      if (tableFamilyDesc == null && replBarrierFamilyDesc == null) {
```

Comment:
Seems 2.0.x had extra table state CF only.  2.1.x Had this replication barrier.  When table state CF is missing it causes the startup issue right?
When its 2.0.x to 2.3.x upgrade, the repBarrier will get auto created?

Reply:
This PR is for 2.3+ releases meaning if we come from HBase 1 to 2.3+, the transition should be seamless. Hence, for this upgrade case, both table and repl_barrier will be missing.

> When its 2.0.x to 2.3.x upgrade, the repBarrier will get auto created?

I have not tried this one. If table CF is handled, I guess repl_barrier too would have been handled. As per @saintstack's testing, the cluster had to come from 1.2 to earlier than 2.3 release and from that release, it went on to 2.3, hence I am assuming HBase 2 (< 2.3) to 2.3 upgrade should have been smooth. @saintstack thoughts?

Reply:
I've run various versions of hbase-2.0-hbase2.2.x upgrades successfully... It was when I tried to go from an hbase1.2 version straight to hbase2.3 that I ran into this issue.

Otherwise, agree w/ your thinking around repl_barrier and table CF.  Is there a case that one might be in place but not the other -- I don't know-- and does the code do right thing(I've not checked)

Reply:
> Is there a case that one might be in place but not the other -- I don't know-- and does the code do right thing(I've not checked)

Exactly, the case where either 'table' or 'repl_barrier' is missing might be weird case and should not happen given that HBase 2 upgrades are not going to face missing CF issues. That's why I kept `&&` here instead of `||` to make sure we are specifically handling HBase 1 to 2.3+ upgrade case.
FYI @anoopsjohn

Reply:
Ok. Good. Thanks.